File: 34.search_for_a_range/searchRange.py
class Solution(object):
    def searchRange(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """
        if not nums:
            return [-1, -1]
        lo = 0
        hi = len(nums)-1
        while lo <= hi:
            if nums[lo] == target and nums[hi] == target:
                return [lo, hi]
            if nums[lo] < target:
                lo += 1
            if nums[hi] > target:
                hi -= 1
        return [-1, -1]

File: 34.search_for_a_range/searchRange.go
package searchRange
// use 2 binary searches to find the first and the last one
func searchRange(nums []int, target int) []int {
	low := 0
	high := len(nums) - 1
	first := -1
	last := -1
	// get the first position
	for low <= high {
			mid := low + (high - low) >> 1
			if nums[mid] == target {
					if mid == 0 || nums[mid-1] < target {
						first = mid
						break
					} else {
						// not the first one
						high = mid - 1
					}
			} else if nums[mid] < target {
				low = mid + 1
			} else {
				high = mid - 1
			}
	}
	low = 0
	high = len(nums) - 1
	// find the last one
	for low <= high {
		mid := low + (high - low) >> 1
		if nums[mid] == target {
			if mid == len(nums) - 1 || nums[mid+1] > target {
				last = mid
				break
			} else {
				low = mid + 1
			}
		} else if nums[mid] > target {
			high = mid -1
		} else {
			low = mid + 1
		}
	}
	res := []int {first, last}
	return res
}